aten/src/ATen/native/transformers/cuda/sdp_utils.cpp

```diff
@@ -303,6 +303,66 @@ bool check_all_tensors_on_device(sdp_params const& params, bool debug) {
 }
 
 } // namespace
+
+static bool check_cudnnshape(sdp_params params, bool debug) {
+  const auto num_heads{params.query.sym_size(1)},
+      query_lengths{params.query.sym_size(2)},
+      head_dim{params.query.sym_size(3)};
+  const bool ok = query_lengths % 64 == 0 && head_dim % 64 == 0;
+  if (debug) {
+    if (!ok) { TORCH_WARN("Shape does not meet cuDNN support constraints (sequence length and head dim divisible by 64)") };
+  }
+  return ok;
+}
+
+static bool check_cudnn_layout(sdp_params params, bool debug) {
+  const int64_t h = params.query.size(1);
+  const int64_t s_q = params.query.size(2);
+  const int64_t d = params.query.size(3);
+  const int64_t s_k = params.key.size(2);
+  const int64_t s_v = params.value.size(2);
+  // corresponds to cuDNN's "packed QKV" layout
+  const bool query_layout_ok = (params.query.stride(0) == s_q * 3 * h * d) &&
+                                 (params.query.stride(1) == d) &&
+                                 (params.query.stride(2) == 3 * h * d) &&
+                                 (params.query.stride(3) == 1);
+  const bool key_layout_ok = (params.key.stride(0) == s_k * 3 * h * d) &&
+                               (params.key.stride(1) == d) &&
+                               (params.key.stride(2) == 3 * h * d) &&
+                               (params.key.stride(3) == 1);
+  const bool value_layout_ok = (params.value.stride(0) == s_v * 3 * h * d) &&
+                                 (params.value.stride(1) == d) &&
+                                 (params.value.stride(2) == 3 * h * d) &&
+                                 (params.value.stride(3) == 1);
+  if (debug) {
+    if (!query_layout_ok) { TORCH_WARN("Query tensor was not in cuDNN-supported packed QKV layout", params.query.strides()); }
+    if (!key_layout_ok) { TORCH_WARN("Key tensor was not in cuDNN-supported packed QKV layout"); }
+    if (!value_layout_ok) { TORCH_WARN("Value tensor was not in cuDNN-supported packed QKV layout"); }
+  }
+  return query_layout_ok && key_layout_ok && value_layout_ok;
+}
+
+static bool check_cudnn_compute_capability(sdp_params params, bool debug) {
+  auto dprops = at::cuda::getCurrentDeviceProperties();
+  return dprops->minor == 0 && dprops->major >= 8;
+}
+
+inline bool can_use_cudnn_attention(sdp_params const& kernel_params, bool print_debug) {
```

Comment:
are all the check being run?

Reply:
These are not being run at the moment as the intention is to revisit them in tandem with backward support when that is done---currently the only real check is to have the environment variable which allows us to definitively check what is actually unsupported vs. just failing a check.